docs/raw/push-notification-server.md

```diff
@@ -0,0 +1,766 @@
+[advertised](#advertising-a-push-notification-server) as described below if no public key filtering is necessary.
+
+## Re-registering with the push notification server
+
+A client SHOULD re-register with the node if the APN or FIREBASE token changes.
+
+When re-registering a client SHOULD ensure that it has the most up-to-date
+`PushNotificationPreferences`, update the part relative to their `installation_id`
+if necessary, increment `version` and send a `PushNotificationRegister` as described above.
+
+Once re-registered, a client SHOULD advertise the changes.
+
+
+## Changing options
+
+This is handled in exactly the same way as re-registering above.
+
+## Unregistering from push notifications
+
+To unregister a client MUST send a `PushNotificationRegister` request as described
+above with `unregister` in `PushNotificationPreferences` set to `true`, or removing
+their device information.
+
+The server MUST remove all data about this user if `unregistering` is `true`,
+apart from the `hash` of the public key and the `version` of the last options,
+in order to make sure that old messages are not processed.
+
+A client MAY unregister from a server on explicit logout if multiple accounts 
+are used on a single device.
+
+## Advertising a push notification server
+
+Each user registered with one or more push notification servers SHOULD
+advertise periodically the push notification services that they have registered with for each device they own.
+
+```protobuf
+message PushNotificationAdvertisementInfo {
+  bytes public_key = 1;
+  string access_token = 2;
+  string installation_id = 3;
+}
+
+message ContactCodeAdvertisement {
+  repeated PushNotificationAdvertisementInfo push_notification_info = 1;
+}
+```
+
+If no filtering is done based on public keys,
+the access token SHOULD be included in the advertisement.
+Otherwise it SHOULD be left empty.
+
+This SHOULD be advertised on the [contact code topic](./10-waku-usage.md#contact-code-topic)
+and SHOULD be coupled with normal contact-code advertisement.
+
+Every time a user register or re-register with a push notification service, their
+contact-code SHOULD be re-advertised.
+
+Multiple servers MAY be advertised for the same `installation_id` for redundancy reasons.
+
+## Discovering a push notification server
+
+[//]: We could query directly the nodes on a shared topic, but for simplicity and 
+bandwidth usage this is the most convenient for now.
+
+To discover a push notification service for a given user, their [contact code topic](./10-waku-usage.md#contact-code-topic)
+SHOULD be listened to.
+A mailserver can be queried for the specific topic to retrieve the most up-to-date
+contact code.
+
+## Querying the push notification server
+
+If a token is not present in the latest advertisement for a user, the server 
+SHOULD be queried directly.
+
+To query a server a message:
+
+```protobuf
+message PushNotificationQuery {
+  repeated bytes public_keys = 1;
+}
```

Comment:
mmh, not sure, first the message is encrypted with the server pk, so to replay this you would need to have access to the private key of the server.

Say that this happen and the key is compromised,an attacker would yes be able to replay the same message, but it would not be able to listen to the reply, as it's encrypted with the pk of the user querying.

If the user has access to both, then it would not make sense to replay a message, as it can just craft a new one.

I mean, we can add a nonce, but in order to replay this effectively you would have to compromise two private keys (or one, the user querying, but at that point you can craft a new message, listen to replies etc), and this message does not change any state, so you would only replay it to get access to information.

Or you think there is still some reason to add a nonce?